app/src/swap/handle_sign_transaction.c

```diff
@@ -0,0 +1,165 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ ********************************************************************************/
+#include "bignum.h"
+#include "crypto.h"
+#include "lib_standard_app/swap_lib_calls.h"
+#include "parser.h"
+#include "swap.h"
+#include "swap_utils.h"
+#include "zxformat.h"
+
+swap_globals_t G_swap_state;
+
+// Save the BSS address where we will write the return value when finished
+static uint8_t *G_swap_sign_return_value_address;
+
+bool copy_transaction_parameters(create_transaction_parameters_t *sign_transaction_params) {
+    if (sign_transaction_params == NULL) {
+        return false;
+    }
+
+    // First copy parameters to stack, and then to global data.
+    // We need this "trick" as the input data position can overlap with globals
+    char destination_address[ADDRESS_MAXSIZE] = {0};
+    uint8_t amount[COIN_AMOUNT_MAXSIZE] = {0};
+    uint8_t amount_length = {0};
+    uint8_t fees[COIN_AMOUNT_MAXSIZE] = {0};
+    uint8_t fees_length = {0};
+    char memo[MEMO_MAXSIZE] = {0};
+
+    // Check and copy destination address
+    if ((destination_address[sizeof(destination_address) - 1] != '\0') || (sign_transaction_params->amount_length > COIN_AMOUNT_MAXSIZE) ||
+        (sign_transaction_params->fee_amount_length > COIN_AMOUNT_MAXSIZE)) {
+        return false;
+    }
+    strncpy(destination_address, sign_transaction_params->destination_address, sizeof(destination_address));
+
+    // Check and copy memo
+    if(strlen(sign_transaction_params->destination_address_extra_id) >= sizeof(G_swap_state.memo)) {
+        return false;
+    }
+    strncpy(memo, sign_transaction_params->destination_address_extra_id, sizeof(G_swap_state.memo));
+
+    // Check and copy amount
+    memcpy(amount, sign_transaction_params->amount, sign_transaction_params->amount_length);
+    amount_length = sign_transaction_params->amount_length;
+
+    // Check and copy fees
+    memcpy(fees, sign_transaction_params->fee_amount, sign_transaction_params->fee_amount_length);
+    fees_length = sign_transaction_params->fee_amount_length;
+
+    // Full reset the global variables
+    os_explicit_zero_BSS_segment();
+
+    // Keep the address at which we'll reply the signing status
+    G_swap_sign_return_value_address = &sign_transaction_params->result;
+
+    // Commit the values read from exchange to the clean global space
+    G_swap_state.amount_length = amount_length;
+    memcpy(G_swap_state.amount, amount, sizeof(amount));
+    G_swap_state.fees_length = fees_length;
+    memcpy(G_swap_state.fees, fees, sizeof(G_swap_state.fees));
+    memcpy(G_swap_state.destination_address, destination_address, sizeof(G_swap_state.destination_address));
```

Comment:
We can't guarantee that the `destination_address` given by exchange fits in `G_swap_state.destination_address`
Check that the resulting address is null terminated